tensorflow/core/graph/mkl_layout_pass.cc

```diff
@@ -2771,6 +2784,59 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
     return n;
   }
 
+  // Find Pad or Conv2D node that can be merged with input node 'm'.
+  // If input 'm' is Pad, then check if there exists Conv2D node that can be
+  // merged with 'm'. If input 'm' is Conv2D, then check if there exists BiasAdd
+  // node that can be merged with 'm'.
+  static Node* GetPadOrConv2D(const Node* m) {
```

Comment:
You are correct, if they are sorted topologically, then we always see Pad before Conv2D. But, in case, the graph is not sorted topologically, and, conv2D is visited before pad, then, the merge will be completed, given that the patterns of the nodes, Pad and Conv2D are mergeable. So, the implementation here is a bit general.

Reply:
Yes, you are right. If nodes are strictly sorted topoligically, then we dont need to check "if m is conv2d find if pad is the source". Here, we have a bit general implementation. This implementation also covers  the case if the graph is not sorted topologically, and conv2d is first visited before pad. But, I understand, this case is rare currently. Please let me know, if this is acceptable. I tried to follow similar convention as https://github.com/NervanaSystems/private-tensorflow/blob/3e2e407c49aa14b525ff6c37538ea4506a152798/tensorflow/core/graph/mkl_layout_pass.cc#L2744